### PR TITLE
feat: compile and run eszip

### PR DESCRIPTION
**Pull request by losfair**
This patch adds [eszip](https://github.com/denoland/eszip) support to the following subcommands:

- `deno compile`
- `deno run`

### CODE REVIEW

File: cli/args/flags.rs [hunk 1]
```diff
@@ -3349,6 +3352,13 @@ fn cached_only_arg() -> Arg {
     .help("Require that remote dependencies are already cached")
 }
 
+fn eszip_arg() -> Arg {
+  Arg::new("eszip")
+    .long("eszip")
+    .action(ArgAction::SetTrue)
+    .help("Run eszip")
```
Comment (ry) [inline cli/args/flags.rs:3359]: hidden(true)

**Review by ry — COMMENTED** [review]
- should have tests
- would be nice if it could be used with denort

File: cli/tools/run/mod.rs [hunk 2]
```diff
@@ -60,13 +69,69 @@ To grant permissions, set them before the script argument. For example:
     deno_dir.upgrade_check_file_path(),
   );
 
-  let main_module = cli_options.resolve_main_module()?;
+  if flags.eszip {
```
Comment (bartlomieju) [inline cli/tools/run/mod.rs:72]: Can you please check that value in `cli/main.rs` in `run_subcommand` and if it's set create a completely new function (eg. `tools::run::run_eszip`). I would prefer we don't interleave this logic with current run subcommand.

File: cli/tools/compile.rs [hunk 3]
```diff
@@ -97,33 +101,73 @@ pub async fn compile(
...
+  let write_result = if compile_flags.eszip {
+    // TODO: write npm vfs
+    let (_, _, node_modules) = binary_writer.pack_node_modules(&mut eszip)?;
+    if let Some(node_modules) = node_modules {
+      eszip.add_opaque_data(
+        "internal://node_modules".to_string(),
+        Arc::from(deno_core::serde_json::to_string(&node_modules)?.as_bytes()),
+      );
+    }
+    file.write_all(&eszip.into_bytes()).map_err(AnyError::from)
```
Comment (ry) [inline cli/tools/compile.rs:127]: This is somewhat abusing the deno compile subcommand to create an eszip - functionality that isn't necessary for SSH. Could you instead move this to the test?

File: cli/tools/compile.rs [hunk 4]
```diff
@@ -97,33 +101,73 @@ pub async fn compile(
...
-  let write_result = binary_writer
-    .write_bin(
-      &mut file,
-      eszip,
-      &module_specifier,
-      &compile_flags,
-      cli_options,
+  let mut temp_filename = output_path.file_name().unwrap().to_owned();
+  temp_filename.push(format!(
+    ".tmp-{}",
```
Comment (ry) [inline cli/tools/compile.rs:106]: This is a race condition bug fix that should probably be landed separately

File: cli/tools/run/mod.rs [hunk 5]
```diff
@@ -60,13 +65,15 @@ To grant permissions, set them before the script argument. For example:
     deno_dir.upgrade_check_file_path(),
   );
 
-  let main_module = cli_options.resolve_main_module()?;
+  if cli_options.unstable_sloppy_imports() {
+    log::warn!(
+      "{} Sloppy imports are not recommended and have a negative impact on performance.",
+      crate::colors::yellow("Warning"),
+    );
+  }
```
Comment (bartlomieju) [inline cli/tools/run/mod.rs:73]: Please revert changes from `run_script` altogether

File: cli/tools/compile.rs [hunk 6]
```diff
@@ -97,33 +101,73 @@ pub async fn compile(
...
+    return Err(err);
   }
 
+  std::fs::rename(&temp_path, &output_path).with_context(|| {
+    format!(
+      "Renaming temporary file '{}' to '{}'",
+      temp_path.display(),
+      output_path.display()
+    )
+  })?;
```
Comment (dsherret) [inline cli/tools/compile.rs:169]: Maybe this should be part of `write_result` so that if the rename fails it will delete the temp file?

**Comment by bartlomieju** [comment]
Closing since we're gonna be upstreaming these changes from another repo.

**Pull request closed without merging**